### ISSUE TITLE
Warn about use of unbuffered os.Signal channel.

### ISSUE THREAD

**Issue by e-dard** [golang/lint#175]
The `os/signal` package will not block when sending a signal, which means that the channel passed to `signal.Notify` should always be buffered.

I think it would be reasonable to add a linter rule that checks to see if the channel passed to `signal.Notify` is buffered or not.

I would be happy to work on this if it would be accepted into the tool.

**Comment by zimmski** [comment]
If you are right that signal.Notify needs a buffere channel no matter what, this should be added as runtime check in the source code anyway. Have you discussed this on the mailing list?

**Comment by e-dard** [comment]
It's not the case that it _need_ a buffered channel no matter what, it's the case that there is no good reason (AFAIK) to pass `signal.Notify` an unbuffered channel. In fact, even the size of the buffer matters if you're listening for more than one type of signal.

Regardless, this is something that can be added as compile-time check, rather than a runtime one, since the size of a channel's buffer cannot be changed at runtime. 

I suppose a question might be whether this is a `vet` check or a `lint` check, but as I understand it `vet` is for situations where the code is likely to lead to incorrect behaviour and `lint` is for stylistic cases. Since an unbuffered channel is unlikely to lead to incorrect behaviour, yet it's technically not the recommended type of buffer to use, I feel a `lint` check would be appropriate.

**Comment by zimmski** [comment]
I do not think that this is stylistic at all. However, the documentation of `signal.Notify` states:

> Package signal will not block sending to c: the caller must ensure that c has sufficient buffer space to keep up with the expected signal rate. For a channel used for notification of just one signal value, a buffer of size 1 is sufficient. 

The last sentence sounds to me that the buffer must be at least one no matter what since being notified for zero messages does not make sense at least to me.

Also, there is an example with the following comment:

> We must use a buffered channel or risk missing the signal if we're not ready to receive when the signal is sent.

That screams "incorrect behaviour" to me. I think that this should be discussed on the mailing list (please CC me), and if the discussion outcome is that "yes at least one but we do not want this as runtime check" I will write the vet check or help you out.
